lib/spack/spack/version/version_types.py

```diff
@@ -61,27 +62,27 @@ def __str__(self):
     def __repr__(self) -> str:
         return f'VersionStrComponent("{self}")'
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
         return isinstance(other, VersionStrComponent) and self.data == other.data
 
-    def __lt__(self, other):
+    def __lt__(self, other) -> bool:
```

Comment:
As above, maybe it's better:
```suggestion
    def __lt__(self, other: object) -> bool:
```
and similarly below in all dunder methods.

Reply:
Not introduced by this PR, so absolutely not a request, but I don't really understand the case of comparing against an `int`. What does it mean:
```console
>>> v3 = VersionStrComponent.from_string("main")
>>> v3.data
4
>>> v3 < 1
False
>>> v3 < 10
False
>>> v3 < 100
False
>>> v2 = VersionStrComponent.from_string("1.2.1")
>>> v2 < 1
True
>>> v2 < 10
True
>>> v2 < 100
True
```
?

Reply:
Ok, I just went back to this. I see why we want `VersionStrComponent` to compare to int, and I see why the use I made in the REPL above is wrong. Anyhow, it feels like this module is missing a lot of docstrings on how it is supposed to behave, given how subtle the comparison is to deal with all the edge cases.

Also, we might want to specify better `Tuple[tuple, tuple]`, which is otherwise not really meaningful in this context. I didn't try it, but I think the right type is something like:
```
VersionComponent = Union[int, VersionStrComponent]
Tuple[Tuple[Tuple[VersionComponent, ...], Tuple[VersionComponent, ...]], Tuple[str]]
```

Reply:
Again, not something introduced by this PR, so we can move this to a following one, but it would be nice to document our internal structure on this part of the code.

Reply:
See above --I want to change as little as possible so I added some `# type: ignore` comments.

Reply:
Fair, as I said this was not something changed in this PR :+1: